cmd/snap-repair/cmd_list.go

```diff
@@ -0,0 +1,183 @@
+
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		fmt.Fprintf(w, "  %s\n", scanner.Text())
+	}
+	if scanner.Err() != nil {
+		fmt.Fprintf(w, "  error: %s\n", scanner.Err())
+	}
+
+}
+
+func showRepairOutput(w io.Writer, issuer, seq, rev string) error {
+	basedir := filepath.Join(dirs.SnapRepairRunDir, issuer, seq)
+	dirents, err := ioutil.ReadDir(basedir)
+	if err != nil {
+		return err
+	}
+	for _, dent := range dirents {
+		name := dent.Name()
+		if strings.HasSuffix(name, ".retry") || strings.HasSuffix(name, ".done") || strings.HasSuffix(name, ".skip") {
+			fmt.Fprintf(w, " output:\n")
+			outputIndented(w, filepath.Join(basedir, name))
+		}
+		if strings.HasSuffix(name, ".script") {
+			fmt.Fprintf(w, " script:\n")
+			outputIndented(w, filepath.Join(basedir, name))
+		}
+	}
+
+	return nil
+}
+
+type repairTrace struct {
+	issuer string
+	seq    string
+	rev    string
+	status string
+}
+
+func newRepairTrace(artifactName, issuerName, seqName, status string) repairTrace {
+	t := repairTrace{
+		issuer: issuerName,
+		seq:    seqName,
+		rev:    "?",
+		status: status,
+	}
+	var rev int
+	if _, err := fmt.Sscanf(artifactName, "r%d.", &rev); err == nil {
+		t.rev = strconv.Itoa(rev)
+	}
+	return t
+}
+
+func (c *cmdList) Execute(args []string) error {
+	w := tabwriter.NewWriter(Stdout, 5, 3, 2, ' ', 0)
+	defer w.Flush()
+
+	// FIXME: this will not currently list the repairs that are
+	//        skipped because of e.g. wrong architecture
+
+	// directory structure is:
+	//  canonical/
+	//    1/
+	//      r0.retry
+	//      r0.script
+	//      r1.done
+	//      r1.script
+	//    2/
+	//      r3.done
+	//      r3.script
+	var repairTraces []repairTrace
+	issuersContent, err := ioutil.ReadDir(dirs.SnapRepairRunDir)
+	if os.IsNotExist(err) {
+		fmt.Fprintf(Stdout, "no repairs yet\n")
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	for _, issuer := range issuersContent {
+		if !issuer.IsDir() {
+			continue
+		}
+		issuerName := issuer.Name()
+
+		seqDir := filepath.Join(dirs.SnapRepairRunDir, issuerName)
+		sequences, err := ioutil.ReadDir(seqDir)
+		if err != nil {
+			continue
+		}
+		for _, seq := range sequences {
+			seqName := seq.Name()
+
+			artifactsDir := filepath.Join(dirs.SnapRepairRunDir, issuerName, seqName)
+			artifacts, err := ioutil.ReadDir(artifactsDir)
+			if err != nil {
+				continue
+			}
+			for _, artifact := range artifacts {
+				artifactName := artifact.Name()
+				switch {
+				case strings.HasSuffix(artifactName, ".retry"):
+					repairTraces = append(repairTraces, newRepairTrace(artifactName, issuerName, seqName, "retry"))
+				case strings.HasSuffix(artifactName, ".skip"):
+					repairTraces = append(repairTraces, newRepairTrace(artifactName, issuerName, seqName, "skip"))
+				case strings.HasSuffix(artifactName, ".done"):
+					repairTraces = append(repairTraces, newRepairTrace(artifactName, issuerName, seqName, "done"))
+				}
+			}
+		}
+	}
+
+	fmt.Fprintf(w, "Issuer\tSeq\tRev\tStatus\n")
```

Comment:
Sorry, I forgot to mention something here: I think it would be nicer if we preserved the original idea of `<issuer>-<id>` as a conventional identity for repairs, at least as far as the user interface goes. These would go into a single "Repair".

Besides being practical to use, this would help conversations, and it also removes the small visual/psyco discomfort that calling this a "sequence" would create when we actually break that sequence (not all machines will see all elements in the sequence).